src/bvh/bvh.rs

```diff
@@ -0,0 +1,401 @@
+    /// Returns the index of the new node in the nodes vector.
+    ///
+    /// [`BVHNode`]: enum.BVHNode.html
+    ///
+    pub fn build<T: BHShape>(shapes: &mut [T],
+                             indices: Vec<usize>,
+                             nodes: &mut Vec<BVHNode>,
+                             parent_index: usize,
+                             depth: u32)
+                             -> usize {
+        // Helper function to accumulate the AABB joint and the centroids AABB
+        fn grow_convex_hull(convex_hull: (AABB, AABB), shape_aabb: &AABB) -> (AABB, AABB) {
+            let center = &shape_aabb.center();
+            let convex_hull_aabbs = &convex_hull.0;
+            let convex_hull_centroids = &convex_hull.1;
+            (convex_hull_aabbs.join(shape_aabb), convex_hull_centroids.grow(center))
+        }
+
+        let mut convex_hull = Default::default();
+        for index in &indices {
+            convex_hull = grow_convex_hull(convex_hull, &shapes[*index].aabb());
+        }
+        let (aabb_bounds, centroid_bounds) = convex_hull;
+
+        // If there is only one element left, don't split anymore
+        if indices.len() == 1 {
+            let shape_index = indices[0];
+            let node_index = nodes.len();
+            nodes.push(BVHNode::Leaf {
+                parent_index: parent_index,
+                depth: depth,
+                shape_index: shape_index,
+            });
+            // Let the shape know the index of the node that represents it
+            shapes[shape_index].set_bh_node_index(node_index);
+            return node_index;
+        }
+
+        // Find the axis along which the shapes are spread the most
+        let split_axis = centroid_bounds.largest_axis();
+        let split_axis_size = centroid_bounds.max[split_axis] - centroid_bounds.min[split_axis];
+
+        /// Defines a Bucket utility object.
+        #[derive(Copy, Clone)]
+        struct Bucket {
+            size: usize,
+            aabb: AABB,
+        }
+
+        impl Bucket {
+            /// Returns an empty bucket.
+            fn empty() -> Bucket {
+                Bucket {
+                    size: 0,
+                    aabb: AABB::empty(),
+                }
+            }
+
+            /// Extends this `Bucket` by the given `AABB`.
+            fn add_aabb(&mut self, aabb: &AABB) {
+                self.size += 1;
+                self.aabb = self.aabb.join(aabb);
+            }
+        }
+
+        /// Joins two `Bucket`s.
+        fn join_bucket(a: Bucket, b: &Bucket) -> Bucket {
+            Bucket {
+                size: a.size + b.size,
+                aabb: a.aabb.join(&b.aabb),
+            }
+        }
+
+        // Create six buckets, and six index assignment vectors
+        const NUM_BUCKETS: usize = 6;
+        let mut buckets = [Bucket::empty(); NUM_BUCKETS];
+        let mut bucket_assignments: [Vec<usize>; NUM_BUCKETS] = Default::default();
```

Comment:
Here we compute the SAH weight of the different split options.
`buckets` stores the bucket definitions of the splits, while
`bucket_assignments` stores the shape indices which were put in this bucket.
This is stored so that we don't have to recompute the índex-partitioning.